programs/solana/programs/ics26-router/src/instructions/recv_packet.rs

```diff
@@ -152,11 +162,17 @@ pub fn recv_packet(ctx: Context<RecvPacket>, msg: MsgRecvPacket) -> Result<()> {
 
     packet_receipt.value = receipt_commitment;
 
-    // TODO: CPI to IBC app's onRecvPacket
-    // For now, we'll create a simple acknowledgement
-    let ack_data = b"packet received".to_vec();
-
-    let acks = vec![ack_data];
+    // CPI to IBC app's onRecvPacket
+    let acknowledgement = on_recv_packet_cpi(
+        &ctx.accounts.ibc_app_program,
+        &ctx.accounts.ibc_app_state,
+        &ctx.accounts.system_program,
+        &msg.packet,
+        &msg.packet.payloads[0],
+        &ctx.accounts.relayer.key(),
+    )?;
+
+    let acks = vec![acknowledgement];
```

Comment:
You need additional validations and processing here. See [this](https://github.com/cosmos/solidity-ibc-eureka/blob/5a19eca83c5440db4e171dccd0811825d554732a/contracts/ICS26Router.sol#L192-L199).

You also don't necessarily want to throw the error here, but instead handle it. (Unless it is a out of gas error, then reverting the whole call is good)